subsys/net/lib/wifi_credentials/wifi_credentials.c

```diff
@@ -0,0 +1,538 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/logging/log.h>
+#include <zephyr/kernel.h>
+#include <zephyr/init.h>
+#include <net/wifi_credentials.h>
+#include <sys/types.h>
+#include <stdlib.h>
+
+LOG_MODULE_REGISTER(wifi_credentials, CONFIG_WIFI_CREDENTIALS_LOG_LEVEL);
+
+#define ENTRY_MAX_LEN sizeof(struct wifi_credentials_personal)
+
+
+static K_MUTEX_DEFINE(wifi_credentials_mutex);
+
+static char configured_ssids[CONFIG_WIFI_CREDENTIALS_MAX_ENTRIES][WIFI_SSID_MAX_LEN];
+
+/**
+ * @brief Stores settings entry in flash
+ *
+ * @param idx credential index
+ * @param buf encoded settings entry
+ * @param buf_len length of buf
+ * @return 0 on success, otherwise a negative error code
+ */
+int wifi_credentials_serialize_entry(size_t idx, const void *buf, size_t buf_len);
+
+/**
+ * @brief Deletes settings entry from flash
+ *
+ * @param idx credential index
+ * @return 0 on success, otherwise a negative error code
+ */
+int wifi_credentials_delete_entry(size_t idx);
+
+/**
+ * @brief Loads settings entry from flash
+ *
+ * @param idx credential index
+ * @param buf encoded settings entry
+ * @param buf_len length of buf
+ * @return 0 on success, otherwise a negative error code
+ */
+int wifi_credentials_deserialize_entry(size_t idx, void *buf, size_t buf_len);
+
+/**
+ * @brief Finds index of given SSID if it exists.
+ *
+ * @param ssid		SSID to look for (buffer of WIFI_SSID_MAX_LEN length)
+ * @return		index if entry is found, -1 otherwise
+ */
+static inline ssize_t lookup_idx(const uint8_t *ssid)
+{
+	for (size_t i = 0; i < CONFIG_WIFI_CREDENTIALS_MAX_ENTRIES; ++i) {
+		if (strcmp(ssid, configured_ssids[i]) == 0) {
+			return i;
+		}
+	}
+
+	return -1;
+}
+
+/**
+ * @brief Finds unused index to store new entry at.
+ *
+ * @return		index if empty slot is found, -1 otherwise
+ */
+static inline ssize_t lookup_unused_idx(void)
+{
+	for (size_t i = 0; i < CONFIG_WIFI_CREDENTIALS_MAX_ENTRIES; ++i) {
+		if (strlen(configured_ssids[i]) == 0) {
+			return i;
+		}
+	}
+
+	return -1;
+}
+
+/**
+ * @brief Load entry into RAM structure.
+ *
+ * @param idx credential index
+ * @param buf encoded settings entry
+ * @param buf_len length of buf
+ */
+static inline void load_entry(size_t idx, const struct wifi_credentials_header *buf)
+{
+	memcpy(configured_ssids[idx], buf->ssid, WIFI_SSID_MAX_LEN);
+}
+
+int wifi_credentials_get_by_ssid_personal_raw(const char *ssid, size_t ssid_len,
+					      struct wifi_credentials_personal *buf)
+{
+	int ret;
+
+	if (ssid == NULL || ssid_len > WIFI_SSID_MAX_LEN || ssid_len == 0) {
+		LOG_ERR("SSID has invalid format");
+		return -EINVAL;
+	}
+	if (buf == NULL) {
+		LOG_ERR("credentials buffer cannot be NULL");
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&wifi_credentials_mutex, K_FOREVER);
+
+	ssize_t idx = lookup_idx(ssid);
+
+	if (idx == -1) {
+		LOG_ERR("entry not found");
```

Comment:
For some logging backends a lot of these logs would be pretty context-less -- I recommend writing more detailed log statements, for instance:

"Cannot retrieve WiFi credentials, no entry found for the provided SSID" in this case

I would make all the other logs more explicit too